### ISSUE TITLE
Slow response when editing CellML files on a network drive 

### ISSUE THREAD

**Issue by WeiweiAi** [opencor/opencor#2457]
OpenCOR Versions: 
Version 0.6 Windows and Snapshot 2021-04-01, Windows 10 (10.0)
OS: 
Windows 10 Pro
Issue: 
CellML model files were saved on UoA home drive (\\files.auckland.ac.nz). I opened the files and edited them. The OpenCOR became extremely slow. It took a long time to switch taps or moving a cursor.
I moved the files to a local drive and the response became smooth.

**Comment by agarny** [comment]
Thanks for the report Weiwei. I have just tried from home on macOS and it's all fine, i.e. no delay in editing/running CellML files.

However, if I try to open a CellML file located on my UoA home drive using my (virtual) Windows machine, then I get an empty file while it's all fine using an editor like Notepad++.

So, there is definitely something funny going on on Windws. :/ I am going to look into it.

**Comment by agarny** [comment]
FTR, on Windows, say that I have the following Windows Explorer instances:
 - `\\files.auckland.ac.nz\myhome` mapped as a network drive (say `X:`); and
 - `\\files.auckland.ac.nz\myhome` (i.e. not mapped).

Then, to open `X:\Models\noble_model_1962.cellml` will be fine (i.e. we can see/edit/etc. the contents of the CellML file without any problems; saving may be considered a bit slow, but I am testing this from home and through the UoA VPN), but if I open `\\files.auckland.ac.nz\myhome\Models\noble_model_1962.cellml` then I end up with an empty file which is considered to be modified while both approaches work fine with Notepad++.

**Comment by agarny** [comment]
@WeiweiAi, could you please confirm the way your filename looks like? Something like `X:\Some\Path\MyModel.cellml` or `\\files.auckland.ac.nz\Some\Path\MyModel.cellml`?

**Comment by agarny** [comment]
FTR, @WeiweiAi using the `X:\` form. If she was to use the canonical form then OpenCOR would show a blank "modified" file. So, she's getting the same behaviour as in my virtual Windows machine. However, in my case I am not getting any lag (be it at home or at the office) while she does. The only difference between our two systems is that she's running a native version of Windows while I am running a virtual version of it, although this shouldn't matter...?

**Comment by agarny** [comment]
@WeiweiAi, I have fixed various issues with files located on a network drive (either in the form of `X:\Some\Path\MyModel.cellml` or `\\files.auckland.ac.nz\Some\Path\MyModel.cellml`). With a bit of luck, it should address the slow response that you are experiencing. So, could you please check this special snapshot and let me know how it goes for you?

It would also be nice if you could try to:
 - run OpenCOR from the command line (after having installed it under `C:\Program Files`; and
 - check that the `Preferences` dialog now/still works fine?

Special snapshot:
 - [OpenCOR-2021-04-21-Windows.exe](https://opencor.ws/OpenCOR-2021-04-21-Windows.exe)
 - [OpenCOR-2021-04-21-Windows.zip](https://opencor.ws/OpenCOR-2021-04-21-Windows.zip)

**Comment by agarny** [comment]
@nickerso, it would be nice if you could also check since @WeiweiAi and you both run Windows natively.

**Comment by nickerso** [comment]
I can confirm that the special snapshot fixes the issue of loading files from the full path, but it seems relative links between files doesn't work. i.e., loading a SED-ML can't find the CellML document (gives a file not found error).

Working with files on a remote server is definitely slower than local files, I'd guess each action that requires the file to be saved will be slow.

**Comment by agarny** [comment]
> I can confirm that the special snapshot fixes the issue of loading files from the full path, but it seems relative links between files doesn't work. i.e., loading a SED-ML can't find the CellML document (gives a file not found error).

Argh, I indeed focussed on CellML files, not SED-ML files. Ok, I am going to try that.

> Working with files on a remote server is definitely slower than local files, I'd guess each action that requires the file to be saved will be slow.

My current theory is that the slowness is/was due to a CellML API call I was doing and which also caused problems with full network paths (see commit c68b12c). @WeiweiAi showed me that switching from editing to simulation mode was very slow. Is that still the case with the special snapshot? If it is then my theory about the CellML API would be wrong. Would be interesting to open, modify and save a (network) file with an editor like Notepad++ and see whether it's also slow.

**Comment by WeiweiAi** [comment]
The snapshot 2021-04-21 still responds very slow to edit files on the network drive. Switching from editing to simulation mode is very slow. In fact, it is slow for any action, like load a CellML file, edit the file, click on any menu tap, etc.

In the cmd prompt, I can run OpenCOR by typing 
C:\Program Files\OpenCOR_test\bin>.\OpenCOR.exe.
The Preferences dialog now/still works fine.

**Comment by agarny** [comment]
> The snapshot 2021-04-21 still responds very slow to edit files on the network drive. Switching from editing to simulation mode is very slow. In fact, it is slow for any action, like load a CellML file, edit the file, click on any menu tap, etc.

Argh! :/

On further investigation, it looks like other people have had slowness issues with network-based files on Windows (using a Qt-based program like OpenCOR). So, it doesn't seem to be an issue specific to OpenCOR. Still, it would be good to have it fixed/addressed, and it would seem that some people have managed to do that. So, I need to investigate things further.

> In the cmd prompt, I can run OpenCOR by typing
> C:\Program Files\OpenCOR_test\bin>.\OpenCOR.exe.
> The Preferences dialog now/still works fine.

Thanks, it's good to know that these things are at least working. :)

Otherwise, on Windows, you don't need to do `.\OpenCOR[.exe]`. You can simply run OpenCOR using `OpenCOR` (or even `opencor` since Windows is not case sensitive).

**Comment by agarny** [comment]
> I can confirm that the special snapshot fixes the issue of loading files from the full path, but it seems relative links between files doesn't work. i.e., loading a SED-ML can't find the CellML document (gives a file not found error).

FWIW, SED-ML files and COMBINE archives work once again (see commit f380de7).

**Comment by agarny** [comment]
Ok, I have generated a new special snapshot which relies on standard C++ (rather than Qt) to load a local file (i.e. on one's computer or on some network drive), as opposed to a remote file (i.e. somewhere on the Internet). Hopefully, that's what was causing the slowness on Windows.

Special snapshot:
 - [OpenCOR-2021-04-27-Windows.exe](https://opencor.ws/OpenCOR-2021-04-27-Windows.exe)
 - [OpenCOR-2021-04-27-Windows.zip](https://opencor.ws/OpenCOR-2021-04-27-Windows.zip)

**Comment by agarny** [comment]
FTR, I tested the new special snapshot with @WeiweiAi and I am not sure whether there is an issue with loading a file (on a network drive) as such. Indeed, say that a network-based file has been loaded and that the `CellML Text` view is active, then to click anywhere in the text editor will be very slow. Same with switching editing views. IOW, as long as a network-based file is open then OpenCOR is very slow (!?).

**Comment by agarny** [comment]
[Here]([OpenCOR-2021-04-27-Windows.exe](https://opencor.ws/application.zip) is a very simple Qt application that can be used to open a text file. Indeed, my current theory is that @WeiweiAi's computer has some network problems and that is a problem with OpenCOR since we try to load a file every second to compute its SHA-1 and determine whether a file has been modified outside of OpenCOR. (A file can also be loaded by OpenCOR when switching views to check whether the file can be rendered by the "new" view.) So, OpenCOR can do quite a few file accesses and if you have some network problems then it's going to slow things down big time.

**Comment by agarny** [comment]
Got confirmation that the small Qt application (`C:\Qt\Examples\Qt-5.12.10\widgets\mainwindows\application`) works fine on @WeiweiAi's computer. So, here is another special snapshot that doesn't regularly check whether a file has been modified.

Special snapshot:
 - [OpenCOR-2021-04-27-Windows.exe](https://opencor.ws/OpenCOR-2021-04-27-Windows.exe)
 - [OpenCOR-2021-04-27-Windows.zip](https://opencor.ws/OpenCOR-2021-04-27-Windows.zip)

**Comment by agarny** [comment]
FTR, the second 2021-04-27 snapshot didn't make any difference on @WeiweiAi's computer.

Note that besides her slow network connection @WeiweiAi had (and which has been addressed by replacing her WiFi connection with a wired connection), it seems like she has an issue with her SSD drive (@WeiweiAi mentioned her laptop taking quite some time to boot while with an SSD drive it ought to be pretty fast). However, I am not sure how and why this would make OpenCOR slow when using network-based files...

So, yes, still unclear what is going on here...

---

The above aside, while working on this issue, I did come across an issue with network-based files on Windows. On Windows, we can map a network drive and reference a network-based file as if it was on a local drive (e.g. `X:\Some\Path\MyModel.cellml`). Alternatively, we can reference a network-based file "natively" (e.g. `\\files.auckland.ac.nz\Some\Path\MyModel.cellml`). In the latter case, OpenCOR didn't handle such a network-based file properly:
 - our `checkFileNameOrUrl()` method considered the file to be a remote file (rather than a "local" file);
 - the CellML API got the XML base of that file wrong; and
 - the CellML API's `loadFromURL()` method doesn't work with "native" network-based files on Windows.
 
So, we created issue #2463.
